File: app/src/main/java/net/yanzm/fabtransformingsample/transition/ImageFade.java
package net.yanzm.fabtransformingsample.transition;

import android.animation.Animator;
import android.animation.AnimatorListenerAdapter;
import android.animation.ValueAnimator;
import android.graphics.drawable.Drawable;
import android.transition.TransitionValues;
import android.transition.Visibility;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;

public class ImageFade extends Visibility {

    public ImageFade() {
    }

    public ImageFade(int mode) {
        setMode(mode);
    }

    @Override
    public Animator onAppear(ViewGroup sceneRoot, final View view, TransitionValues startValues,
                             TransitionValues endValues) {

        view.setAlpha(0f);

        if (!(view instanceof ImageView)) {
            return null;
        }

        final ImageView iv = (ImageView) view;
        final Drawable drawable = iv.getDrawable();

        if (drawable == null) {
            return null;
        }

        drawable.setAlpha(0);
        final ValueAnimator anim = ValueAnimator.ofInt(0, 255);
        anim.addUpdateListener(new ValueAnimator.AnimatorUpdateListener() {
            @Override
            public void onAnimationUpdate(ValueAnimator valueAnimator) {
                final int alpha = (int) valueAnimator.getAnimatedValue();
                drawable.setAlpha(alpha);
            }
        });
        anim.addListener(new AnimatorListenerAdapter() {
            @Override
            public void onAnimationStart(Animator animation) {
                super.onAnimationStart(animation);
                view.setAlpha(1f);
            }
        });
        return anim;
    }

    @Override
    public Animator onDisappear(ViewGroup sceneRoot, final View view, TransitionValues startValues,
                                TransitionValues endValues) {
        if (!(view instanceof ImageView)) {
            return null;
        }

        final ImageView iv = (ImageView) view;
        final Drawable drawable = iv.getDrawable();

        if (drawable == null) {
            return null;
        }

        final ValueAnimator anim = ValueAnimator.ofInt(255, 0);
        anim.addUpdateListener(new ValueAnimator.AnimatorUpdateListener() {
            @Override
            public void onAnimationUpdate(ValueAnimator valueAnimator) {
                final int alpha = (int) valueAnimator.getAnimatedValue();
                drawable.setAlpha(alpha);
            }
        });
        return anim;
    }
}

File: app/src/main/java/net/yanzm/fabtransformingsample/transition/ImageFade.kt
package net.yanzm.fabtransformingsample.transition

import android.animation.Animator
import android.animation.AnimatorListenerAdapter
import android.animation.ValueAnimator
import android.transition.TransitionValues
import android.transition.Visibility
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView

class ImageFade(mode: Int) : Visibility() {

    init {
        setMode(mode)
    }

    override fun onAppear(
        sceneRoot: ViewGroup,
        view: View,
        startValues: TransitionValues,
        endValues: TransitionValues
    ): Animator? {
        val imageView = view as? ImageView ?: return null
        val drawable = imageView.drawable ?: return null
        imageView.alpha = 0f
        drawable.alpha = 0
        return ValueAnimator.ofInt(0, 255).apply {
            addUpdateListener {
                drawable.alpha = it.animatedValue as Int
            }
            addListener(object : AnimatorListenerAdapter() {
                override fun onAnimationStart(animation: Animator) {
                    super.onAnimationStart(animation)
                    view.setAlpha(1f)
                }
            })
        }
    }

    override fun onDisappear(
        sceneRoot: ViewGroup,
        view: View,
        startValues: TransitionValues,
        endValues: TransitionValues
    ): Animator? {
        val imageView = view as? ImageView ?: return null
        val drawable = imageView.drawable ?: return null
        return ValueAnimator.ofInt(255, 0).apply {
            addUpdateListener {
                drawable.alpha = it.animatedValue as Int
            }
        }
    }
}
